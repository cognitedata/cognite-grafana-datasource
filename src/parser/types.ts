import { STSClientFilter } from './ts';

export const FilterTypes = {
  RegexNotEquals: '!~' as FilterType,
  RegexEquals: '=~' as FilterType,
  NotEquals: '!=' as FilterType,
  Equals: '=' as FilterType,
};
export type FilterType = '!~' | '=~' | '!=' | '=';

export enum QueryParserTypes {
  assets = 'assets',
  events = 'events',
}

export interface QueryParserType {
  type: QueryParserTypes;
}

export interface QueryParserResponse extends QueryParserType {
  query: QueryCondition[];
}

export type ParsedFilter = STSClientFilter;

export interface QueryCondition {
  [key: string]: any;
}

export interface ParserResponse extends QueryParserType {
  params: QueryCondition;
  filters: ParsedFilter[];
}
