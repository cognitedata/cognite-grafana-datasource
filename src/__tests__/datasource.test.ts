import { cloneDeep } from 'lodash';
import { getMockedDataSource, getDataqueryResponse, getItemsResponseObject } from './utils';
import { Tab, InputQueryTarget } from '../types';
import ms from 'ms';

jest.mock('grafana/app/core/utils/datemath');
jest.mock('../cache');

type QueryTargetLike = Partial<InputQueryTarget>;

const { ds, backendSrvMock, templateSrvMock } = getMockedDataSource();

const tsError = {
  status: 400,
  data: {
    error: {
      code: 400,
      message: 'error message',
    },
  },
};
const externalIdPrefix = 'Timeseries';

describe('Datasource Query', () => {
  const options: any = {
    targets: [],
    range: {
      from: '1549336675000',
      to: '1549338475000',
    },
    interval: '30s',
    intervalMs: ms('30s'),
    maxDataPoints: 760,
    format: 'json',
    panelId: 1,
    dashboardId: 1,
  };

  beforeEach(() => {
    options.panelId++;
  });

  describe('Given an older queryTarget format', () => {
    let results;
    const id = 123;
    const aggregates = ['average'];
    const items = [{ id }];
    const oldTarget: QueryTargetLike = {
      aggregation: aggregates[0],
      target: id,
      refId: 'A',
    };
    const [start, end] = [+options.range.from, +options.range.to];

    it('should generate the correct query', async () => {
      backendSrvMock.datasourceRequest = jest
        .fn()
        .mockImplementation(x => Promise.resolve(getDataqueryResponse(x.data)));
      results = await ds.fetchTimeseriesForTargets([oldTarget] as any, options);

      expect(backendSrvMock.datasourceRequest).toBeCalledTimes(1);
      expect(backendSrvMock.datasourceRequest.mock.calls[0][0].data).toEqual({
        end,
        start,
        items,
        aggregates,
        limit: 10000,
        granularity: '30s',
      });
    });

    it('should give correct meta responses', async () => {
      expect(results.succeded[0].metadata).toEqual({
        target: oldTarget,
        labels: [''],
      });
      expect(results.succeded[0].result).toEqual(getDataqueryResponse({ items, aggregates }));
    });
  });

  describe('Given "Select Timeseries" queries', () => {
    let result;
    const id = 789;
    const tsTargetA: QueryTargetLike = {
      aggregation: 'none',
      refId: 'A',
      target: 123,
      tab: Tab.Timeseries,
    };
    const tsTargetB: QueryTargetLike = {
      ...tsTargetA,
      aggregation: 'count',
      refId: 'B',
      target: 456,
      granularity: '20m',
    };
    const tsTargetC: QueryTargetLike = {
      ...tsTargetA,
      aggregation: 'step',
      refId: 'C',
      target: id,
      label: '{{description}}-{{id}}',
    };

    const tsResponse = getItemsResponseObject([
      {
        id,
        description: 'test timeseries',
      },
    ]);

    beforeAll(async () => {
      options.intervalMs = 1;
      options.targets = [tsTargetA, tsTargetB, tsTargetC];
      backendSrvMock.datasourceRequest = jest
        .fn()
        .mockImplementationOnce(() => Promise.resolve(tsResponse))
        .mockImplementation(x => Promise.resolve(getDataqueryResponse(x.data, externalIdPrefix)));
      result = await ds.query(options);
    });

    it('should generate the correct queries', () => {
      expect(backendSrvMock.datasourceRequest).toBeCalledTimes(4);
      expect(backendSrvMock.datasourceRequest.mock.calls[0][0]).toMatchSnapshot();
      expect(backendSrvMock.datasourceRequest.mock.calls[1][0]).toMatchSnapshot();
      expect(backendSrvMock.datasourceRequest.mock.calls[2][0]).toMatchSnapshot();
      expect(backendSrvMock.datasourceRequest.mock.calls[3][0]).toMatchSnapshot();
    });
    it('should return correct datapoints and labels', () => {
      const { target: targetA } = tsTargetA;
      const { target: targetB, aggregation: aggregationB } = tsTargetB;
      const {
        data: {
          items: [{ id: targetC, description }],
        },
      } = tsResponse;

      expect(result.data[0].target).toEqual(`${externalIdPrefix}${targetA}`);
      expect(result.data[1].target).toEqual(`${aggregationB} ${externalIdPrefix}${targetB}`);
      expect(result.data[2].target).toEqual(`${description}-${targetC}`);
    });
  });

  describe('Given "Select Timeseries" queries with errors', () => {
    let result;
    const id = 123;
    const tsTargetA: QueryTargetLike = {
      aggregation: 'none',
      refId: 'A',
      target: id,
      tab: Tab.Timeseries,
    };
    const tsTargetB: QueryTargetLike = {
      ...tsTargetA,
      refId: 'B',
    };

    beforeAll(async () => {
      options.intervalMs = ms('1m');
      options.targets = [tsTargetA, tsTargetB];
      backendSrvMock.datasourceRequest = jest
        .fn()
        .mockRejectedValueOnce(tsError)
        .mockRejectedValueOnce({});
      result = await ds.query(options);
    });

    it('should generate the correct query', () => {
      expect(backendSrvMock.datasourceRequest).toBeCalledTimes(2);
      expect(backendSrvMock.datasourceRequest.mock.calls[0][0]).toMatchSnapshot();
      expect(backendSrvMock.datasourceRequest.mock.calls[1][0]).toMatchSnapshot();
    });
    it('should return an empty array', () => {
      expect(result).toEqual({ data: [] });
    });
    it('should display errors for malformed queries', () => {
      expect(tsTargetA.error).toBeDefined();
      expect(tsTargetA.error).not.toHaveLength(0);
      expect(tsTargetB.error).toBeDefined();
      expect(tsTargetB.error).not.toHaveLength(0);
    });
  });

  describe('Given "Select Timeseries from Asset" queries', () => {
    let result;
    const targetA: QueryTargetLike = {
      aggregation: 'none',
      refId: 'A',
      tab: Tab.Asset,
      assetQuery: {
        target: '123',
        timeseries: [],
        includeSubtrees: false,
      },
    };
    const targetB: QueryTargetLike = {
      ...targetA,
      aggregation: 'min',
      refId: 'B',
      target: '',
      granularity: '20m',
      assetQuery: {
        target: '456',
        timeseries: [],
        includeSubtrees: true,
      },
    };
    const targetC: QueryTargetLike = {
      ...targetA,
      aggregation: 'max',
      refId: 'C',
      target: '',
      label: '{{description}}-{{externalId}}',
      assetQuery: {
        target: '789',
        timeseries: [],
        includeSubtrees: false,
        func: 'should not be evaluated',
      },
    };
    const targetD: QueryTargetLike = {
      ...targetA,
      aggregation: 'tv',
      refId: 'D',
      target: '',
      label: '{{description}}',
      assetQuery: {
        target: '[[AssetVariable]]',
        timeseries: [],
        includeSubtrees: false,
        func: 'should not be evaluated',
      },
    };
    const targetEmpty: QueryTargetLike = {
      ...targetA,
      aggregation: 'average',
      refId: 'E',
      target: 123,
      label: '{{description}}-{{externalId}}',
      assetQuery: {
        target: '000',
        timeseries: [],
        includeSubtrees: true,
      },
    };
    const targetError1: QueryTargetLike = {
      ...cloneDeep(targetA),
      refId: 'E',
    };
    const targetError2: QueryTargetLike = {
      ...cloneDeep(targetA),
      refId: 'F',
    };

    const tsResponseA = getItemsResponseObject([
      { id: 123, externalId: 'Timeseries123', description: 'test timeseries' },
    ]);
    const tsResponseB = getItemsResponseObject([
      { id: 123, externalId: 'Timeseries123', description: 'test timeseries' },
      { id: 456, externalId: 'Timeseries456', description: 'test timeseries' },
    ]);
    const tsResponseC = getItemsResponseObject([
      { id: 123, externalId: 'Timeseries123', description: 'test timeseriesA' },
      { id: 456, externalId: 'Timeseries456', description: 'test timeseriesB' },
      { id: 789, externalId: 'Timeseries789', description: 'test timeseriesC' },
    ]);
    const tsResponseEmpty = getItemsResponseObject([]);

    beforeAll(async () => {
      options.intervalMs = ms('6m');
      options.targets = [
        targetA,
        targetB,
        targetC,
        targetD,
        targetEmpty,
        targetError1,
        targetError2,
      ];
      backendSrvMock.datasourceRequest = jest
        .fn()
        .mockImplementationOnce(() => Promise.resolve(tsResponseA))
        .mockImplementationOnce(() => Promise.resolve(tsResponseB))
        .mockImplementationOnce(() => Promise.resolve(tsResponseC))
        .mockImplementationOnce(() => Promise.resolve(tsResponseA))
        .mockImplementationOnce(() => Promise.resolve(tsResponseEmpty))
        .mockRejectedValueOnce(tsError)
        .mockRejectedValueOnce({})
        .mockImplementation(x => Promise.resolve(getDataqueryResponse(x.data, externalIdPrefix)));
      result = await ds.query(options);
    });
    afterAll(() => {
      templateSrvMock.replace.mockClear();
    });

    it('should generate the correct queries', () => {
      expect(backendSrvMock.datasourceRequest).toBeCalledTimes(11);
      for (let i = 0; i < backendSrvMock.datasourceRequest.mock.calls.length; ++i) {
        expect(backendSrvMock.datasourceRequest.mock.calls[i][0]).toMatchSnapshot();
      }
    });

    it('should replace filter assetQuery [[variable]] with its value', () => {
      expect(backendSrvMock.datasourceRequest.mock.calls[3][0].data.filter.assetIds[0]).toEqual(
        '123'
      );
    });

    it('should return correct datapoints and labels', () => {
      expect(result).toMatchSnapshot();
    });

    it('should display errors for malformed queries', () => {
      expect(targetError1.error).toBe('[400 ERROR] error message');
      expect(targetError2.error).toBe('Unknown error');
    });
  });

  describe('Given custom queries', () => {
    let result;
    const targetA: QueryTargetLike = {
      aggregation: 'none',
      refId: 'A',
      target: 123,
      tab: Tab.Custom,
      expr: 'ts{}',
    };
    const targetB: QueryTargetLike = {
      ...cloneDeep(targetA),
      expr:
        "ts{description!='test timeseriesC', metadata={key1='value1', key2!~'.*2'}, aggregate='discreteVariance', granularity='10d'}",
      refId: 'B',
    };
    const targetC: QueryTargetLike = {
      ...cloneDeep(targetA),
      expr:
        "ts{name='Timeseries1', description!='test timeseriesC', metadata={key1='value1', key2!~'.*2'}, aggregate='discreteVariance'}",
      refId: 'C',
    };
    const targetD: QueryTargetLike = {
      ...cloneDeep(targetA),
      expr: 'ts{externalId="[[TimeseriesVariable]]"}',
      refId: 'D',
    };
    const targetE: QueryTargetLike = {
      ...cloneDeep(targetA),
      assetQuery: {
        ...targetA.assetQuery,
        target: '$AssetVariable'
      },
      expr: 'ts{externalId!="$TimeseriesVariable"}',
      label: '{{description}} {{metadata.key1}}',
      refId: 'E',
    };
    const targetF: QueryTargetLike = {
      ...cloneDeep(targetA),
      refId: 'F',
      expr: 'ts{externalId=~".*}',
    };
    const targetG: QueryTargetLike = {
      ...cloneDeep(targetA),
      refId: 'G',
      expr: 'ts{externalId=~".*"}[avg',
    };
    const targetH: QueryTargetLike = {
      ...cloneDeep(targetA),
      refId: 'H',
      expr: 'ts{externalId=~".*"',
    };
    const targetI: QueryTargetLike = {
      ...cloneDeep(targetA),
      refId: 'I',
      expr: '',
    };
    const targetJ: QueryTargetLike = {
      ...cloneDeep(targetA),
      refId: 'J',
      expr: '-',
    };
    const tsResponse = getItemsResponseObject([
      {
        id: 1,
        externalId: 'Timeseries1',
        description: 'test timeseriesA',
        metadata: { key1: 'value1', key2: 'value3' },
      },
      {
        id: 2,
        externalId: 'Timeseries2',
        description: 'test timeseriesB',
        metadata: { key1: 'value1', key2: 'value2' },
      },
      {
        id: 3,
        externalId: 'Timeseries3',
        description: 'test timeseriesC',
        metadata: { key1: 'value1' },
      }
    ]);

    beforeAll(async () => {
      options.intervalMs = ms('1d');
      options.targets = [
        targetA,
        targetB,
        targetC,
        targetD,
        targetE,
        targetF,
        targetG,
        targetH,
        targetI,
        targetJ,
      ];
      const listMock = async () => {
        return cloneDeep(tsResponse);
      }
      const dataMock = async x => {
        return getDataqueryResponse(x.data, externalIdPrefix, 0)
      }
      backendSrvMock.datasourceRequest = jest.fn()
        .mockImplementationOnce(listMock)
        .mockImplementationOnce(listMock)
        .mockImplementationOnce(listMock)
        .mockImplementationOnce(listMock)
        .mockImplementationOnce(listMock)
        .mockImplementationOnce(listMock)
        .mockImplementationOnce(listMock)
        .mockImplementationOnce(listMock)
        .mockImplementationOnce(dataMock)
        .mockImplementationOnce(dataMock)
        .mockImplementationOnce(dataMock)
        .mockImplementationOnce(dataMock)
     
      result = await ds.query(options);
    });
    afterAll(() => {
      templateSrvMock.replace.mockClear();
    });

    it('should generate the correct filtered queries', () => {
      expect(backendSrvMock.datasourceRequest).toBeCalledTimes(14);
      for (let i = 0; i < backendSrvMock.datasourceRequest.mock.calls.length; ++i) {
        expect(backendSrvMock.datasourceRequest.mock.calls[i][0]).toMatchSnapshot();
      }
    });

    it('should return correct datapoints and labels', () => {
      expect(result).toMatchSnapshot();
    });

    it('should display errors for malformed queries', () => {
      expect(targetF.error).toBeDefined();
      expect(targetF.error).not.toHaveLength(0);
      expect(targetG.error).toBeDefined();
      expect(targetG.error).not.toHaveLength(0);
      expect(targetH.error).toBeDefined();
      expect(targetH.error).not.toHaveLength(0);
      expect(targetI.error).toBeDefined();
      expect(targetJ.error).toBeDefined();
    });
  });

  describe('Given custom queries with functions', () => {
    let result;
    const targetA: QueryTargetLike = {
      aggregation: 'none',
      refId: 'A',
      target: 123,
      tab: Tab.Custom,
      expr: 'ts{} + pi()',
    };
    const targetB: QueryTargetLike = {
      ...cloneDeep(targetA),
      refId: 'B',
      expr: '       ts{} + ts{id=1}   ',
    };
    const targetC: QueryTargetLike = {
      ...cloneDeep(targetA),
      refId: 'C',
      expr: 'ts{} * ts{aggregate="average"} - ts{aggregate="average", granularity="10s"}',
      label: '{{description}} {{metadata.key1}}',
    };
    const targetD: QueryTargetLike = {
      ...cloneDeep(targetA),
      refId: 'D',
      expr:
        'ts{} * ts{externalId="[[TimeseriesVariable]]", aggregate="average"} - ts{externalId="[[TimeseriesVariable]]", aggregate="average", granularity="10m"}',
    };
    const targetE: QueryTargetLike = {
      ...cloneDeep(targetA),
      refId: 'E',
      expr: 'ts{asdaklj}',
    };
    const targetF: QueryTargetLike = {
      ...cloneDeep(targetA),
      refId: 'F',
      expr: 'no',
    };
    const targetG: QueryTargetLike = {
      ...cloneDeep(targetA),
      refId: 'G',
      expr: 'sum(ts{})',
      label: 'sum',
    };
    const targetH: QueryTargetLike = {
      ...cloneDeep(targetA),
      refId: 'H',
      expr:
        'sum(ts{aggregate="average"}) + sum(ts{aggregate="average", granularity="1h"}) * max(ts{aggregate="count"})/min(ts{externalId="Timeseries1"}) - avg(ts{aggregate="average"}) - 3*ts{}',
    };
    const targetI: QueryTargetLike = {
      ...cloneDeep(targetA),
      refId: 'I',
      expr: 'max(max(ts{},5),5) + max(ts{})',
    };
    const targetJ: QueryTargetLike = {
      ...cloneDeep(targetA),
      refId: 'J',
      expr: 'ts{externalId="[[TimeseriesVariable]]"}',
      label: '{{description}} : [[TimeseriesVariable]]',
    };

    const tsResponse = getItemsResponseObject([
      {
        externalId: 'Timeseries1',
        id: 1,
        description: 'test timeseriesA',
        metadata: { key1: 'value1', key2: 'value3' },
      },
      {
        externalId: 'Timeseries2',
        id: 2,
        description: 'test timeseriesB',
        metadata: { key1: 'value1', key2: 'value2' },
      },
      {
        externalId: 'Timeseries3',
        id: 3,
        description: 'test timeseriesC',
        metadata: { key1: 'value1' },
      },
    ]);

    beforeAll(async () => {
      options.intervalMs = ms('2.5h');
      options.targets = [
        targetA,
        targetB,
        targetC,
        targetD,
        targetE,
        targetF,
        targetG,
        targetH,
        targetI,
        targetJ,
      ];
      backendSrvMock.datasourceRequest = jest
        .fn()
        .mockImplementation(async (data: any) => {
          if (data.url.includes('/byids') || data.url.includes('timeseries/list')) {
            return cloneDeep(tsResponse)
          }  if (data.url.includes('/synthetic/query')) {
            return getDataqueryResponse(data.data, externalIdPrefix, 0)
          }
          throw new Error('no mock')
        })
      result = await ds.query(options);
    });

    it('should generate the correct filtered queries', () => {
      expect(backendSrvMock.datasourceRequest).toBeCalledTimes(30);
      for (let i = 0; i < backendSrvMock.datasourceRequest.mock.calls.length; ++i) {
        expect(backendSrvMock.datasourceRequest.mock.calls[i][0]).toMatchSnapshot();
      }
    });

    it('should return correct datapoints and labels', () => {
      expect(result).toMatchSnapshot();
    });

    it('should display errors for malformed queries', () => {
      expect(targetE.error).toBeDefined();
      expect(targetE.error).not.toHaveLength(0);
      expect(targetF.error).toBeDefined();
      expect(targetF.error).not.toHaveLength(0);
    });
  });
  describe('Given multiple "Select Timeseries from Asset" queries in a row', () => {
    const results = [];
    const targetA: QueryTargetLike = {
      aggregation: 'none',
      refId: 'A',
      target: 123,
      tab: Tab.Asset,
      assetQuery: {
        target: '123',
        timeseries: [],
        includeSubtrees: false,
      },
    };
    const targetB: QueryTargetLike = {
      ...targetA,
      assetQuery: {
        target: '123',
        timeseries: [],
        includeSubtrees: true,
      },
    };
    const targetC: QueryTargetLike = {
      ...targetA,
      assetQuery: {
        target: '456',
        timeseries: [],
        includeSubtrees: true,
      },
    };

    const tsResponseA = getItemsResponseObject([
      { id: 123, externalId: 'Timeseries123', description: 'test timeseries' },
    ]);
    const tsResponseB = getItemsResponseObject([
      { id: 123, externalId: 'Timeseries123', description: 'test timeseries' },
      { id: 456, externalId: 'Timeseries456', description: 'test timeseries' },
    ]);
    const tsResponseEmpty = getItemsResponseObject([]);

    beforeAll(async () => {
      options.intervalMs = ms('6m');
      options.targets = [targetA];
      backendSrvMock.datasourceRequest = jest
        .fn()
        .mockImplementationOnce(() => Promise.resolve(tsResponseEmpty))
        .mockImplementationOnce(() => Promise.resolve(tsResponseA))
        .mockImplementationOnce(x =>
          Promise.resolve(getDataqueryResponse(x.data, externalIdPrefix))
        )
        .mockImplementationOnce(() => Promise.resolve(tsResponseB))
        .mockImplementation(x => Promise.resolve(getDataqueryResponse(x.data, externalIdPrefix)));
      results.push(await ds.query(options));
      results.push(await ds.query(options));
      options.targets = [targetB];
      results.push(await ds.query(options));
      options.targets = [targetC];
      results.push(await ds.query(options));
    });

    it('should generate the correct queries and not requery for asset timeseries', () => {
      expect(backendSrvMock.datasourceRequest).toBeCalledTimes(5);
      for (let i = 0; i < backendSrvMock.datasourceRequest.mock.calls.length; ++i) {
        expect(backendSrvMock.datasourceRequest.mock.calls[i][0]).toMatchSnapshot();
      }
    });
    it('should return correct datapoints and labels', () => {
      for (const result of results) {
        expect(result).toMatchSnapshot();
      }
    });
  });
});
