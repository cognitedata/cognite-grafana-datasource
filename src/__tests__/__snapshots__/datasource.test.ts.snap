// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Datasource Query Given "Select Timeseries from Asset" queries should generate the correct queries 1`] = `
Object {
  "data": Object {
    "filter": Object {
      "assetIds": Array [
        "789",
      ],
    },
    "limit": 101,
  },
  "method": "POST",
  "url": "/api/datasources/proxy/6/cogniteapi/TestProject/timeseries/list",
}
`;

exports[`Datasource Query Given "Select Timeseries from Asset" queries should generate the correct queries 2`] = `
Object {
  "data": Object {
    "filter": Object {
      "assetIds": Array [
        "123",
      ],
    },
    "limit": 101,
  },
  "method": "POST",
  "url": "/api/datasources/proxy/6/cogniteapi/TestProject/timeseries/list",
}
`;

exports[`Datasource Query Given "Select Timeseries from Asset" queries should generate the correct queries 3`] = `
Object {
  "data": Object {
    "filter": Object {
      "assetIds": Array [
        "1",
      ],
    },
    "limit": 101,
  },
  "method": "POST",
  "url": "/api/datasources/proxy/6/cogniteapi/TestProject/timeseries/list",
}
`;

exports[`Datasource Query Given "Select Timeseries from Asset" queries should generate the correct queries 4`] = `
Object {
  "data": Object {
    "filter": Object {
      "assetIds": Array [
        "2",
      ],
    },
    "limit": 101,
  },
  "method": "POST",
  "url": "/api/datasources/proxy/6/cogniteapi/TestProject/timeseries/list",
}
`;

exports[`Datasource Query Given "Select Timeseries from Asset" queries should generate the correct queries 5`] = `
Object {
  "data": Object {
    "items": Array [
      Object {
        "id": 123,
      },
      Object {
        "id": 456,
      },
      Object {
        "id": 789,
      },
    ],
  },
  "method": "POST",
  "url": "/api/datasources/proxy/6/cogniteapi/TestProject/timeseries/byids",
}
`;

exports[`Datasource Query Given "Select Timeseries from Asset" queries should generate the correct queries 6`] = `
Object {
  "data": Object {
    "items": Array [
      Object {
        "id": 123,
      },
    ],
  },
  "method": "POST",
  "url": "/api/datasources/proxy/6/cogniteapi/TestProject/timeseries/byids",
}
`;

exports[`Datasource Query Given "Select Timeseries from Asset" queries should generate the correct queries 7`] = `
Object {
  "data": Object {
    "aggregates": Array [
      "max",
    ],
    "end": 1549338475000,
    "granularity": "6m",
    "items": Array [
      Object {
        "id": 123,
      },
      Object {
        "id": 456,
      },
      Object {
        "id": 789,
      },
    ],
    "limit": 3333,
    "start": 1549336675000,
  },
  "method": "POST",
  "requestId": "1_7_C",
  "url": "/api/datasources/proxy/6/cogniteapi/TestProject/timeseries/data/list",
}
`;

exports[`Datasource Query Given "Select Timeseries from Asset" queries should generate the correct queries 8`] = `
Object {
  "data": Object {
    "aggregates": Array [
      "tv",
    ],
    "end": 1549338475000,
    "granularity": "6m",
    "items": Array [
      Object {
        "id": 123,
      },
    ],
    "limit": 10000,
    "start": 1549336675000,
  },
  "method": "POST",
  "requestId": "1_7_D",
  "url": "/api/datasources/proxy/6/cogniteapi/TestProject/timeseries/data/list",
}
`;

exports[`Datasource Query Given "Select Timeseries from Asset" queries should return correct datapoints and labels 1`] = `
Object {
  "data": Array [
    Object {
      "datapoints": Array [
        Array [
          0,
          1549336675000,
        ],
        Array [
          1,
          1549337275000,
        ],
        Array [
          2,
          1549337875000,
        ],
        Array [
          3,
          1549338475000,
        ],
      ],
      "target": "test timeseriesA-Timeseries123",
    },
    Object {
      "datapoints": Array [
        Array [
          0,
          1549336675000,
        ],
        Array [
          1,
          1549337275000,
        ],
        Array [
          2,
          1549337875000,
        ],
        Array [
          3,
          1549338475000,
        ],
      ],
      "target": "test timeseriesB-Timeseries456",
    },
    Object {
      "datapoints": Array [
        Array [
          0,
          1549336675000,
        ],
        Array [
          1,
          1549337275000,
        ],
        Array [
          2,
          1549337875000,
        ],
        Array [
          3,
          1549338475000,
        ],
      ],
      "target": "test timeseriesC-Timeseries789",
    },
    Object {
      "datapoints": Array [
        Array [
          0,
          1549336675000,
        ],
        Array [
          1,
          1549337275000,
        ],
        Array [
          2,
          1549337875000,
        ],
        Array [
          3,
          1549338475000,
        ],
      ],
      "target": "test timeseries",
    },
  ],
}
`;

exports[`Datasource Query Given "Select Timeseries" queries should generate the correct queries 1`] = `
Object {
  "data": Object {
    "items": Array [
      Object {
        "id": 789,
      },
    ],
  },
  "method": "POST",
  "url": "/api/datasources/proxy/6/cogniteapi/TestProject/timeseries/byids",
}
`;

exports[`Datasource Query Given "Select Timeseries" queries should generate the correct queries 2`] = `
Object {
  "data": Object {
    "end": 1549338475000,
    "items": Array [
      Object {
        "id": 123,
      },
    ],
    "limit": 100000,
    "start": 1549336675000,
  },
  "method": "POST",
  "requestId": "1_3_A",
  "url": "/api/datasources/proxy/6/cogniteapi/TestProject/timeseries/data/list",
}
`;

exports[`Datasource Query Given "Select Timeseries" queries should generate the correct queries 3`] = `
Object {
  "data": Object {
    "aggregates": Array [
      "count",
    ],
    "end": 1549338475000,
    "granularity": "20m",
    "items": Array [
      Object {
        "id": 456,
      },
    ],
    "limit": 10000,
    "start": 1549336675000,
  },
  "method": "POST",
  "requestId": "1_3_B",
  "url": "/api/datasources/proxy/6/cogniteapi/TestProject/timeseries/data/list",
}
`;

exports[`Datasource Query Given "Select Timeseries" queries should generate the correct queries 4`] = `
Object {
  "data": Object {
    "aggregates": Array [
      "step",
    ],
    "end": 1549338475000,
    "granularity": "1s",
    "items": Array [
      Object {
        "id": 789,
      },
    ],
    "limit": 10000,
    "start": 1549336675000,
  },
  "method": "POST",
  "requestId": "1_3_C",
  "url": "/api/datasources/proxy/6/cogniteapi/TestProject/timeseries/data/list",
}
`;

exports[`Datasource Query Given custom queries should generate the correct filtered queries 1`] = `
Object {
  "data": Object {
    "filter": Object {
      "metadata": Object {
        "key1": "value1",
      },
    },
    "limit": 1000,
  },
  "method": "POST",
  "url": "/api/datasources/proxy/6/cogniteapi/TestProject/timeseries/list",
}
`;

exports[`Datasource Query Given custom queries should generate the correct filtered queries 2`] = `
Object {
  "data": Object {
    "items": Array [
      Object {
        "id": 1,
      },
    ],
  },
  "method": "POST",
  "url": "/api/datasources/proxy/6/cogniteapi/TestProject/timeseries/byids",
}
`;

exports[`Datasource Query Given custom queries should generate the correct filtered queries 3`] = `
Object {
  "data": Object {
    "items": Array [
      Object {
        "externalId": "Timeseries1",
      },
    ],
  },
  "method": "POST",
  "url": "/api/datasources/proxy/6/cogniteapi/TestProject/timeseries/byids",
}
`;

exports[`Datasource Query Given custom queries should generate the correct filtered queries 4`] = `
Object {
  "data": Object {
    "items": Array [
      Object {
        "end": 1549338475000,
        "expression": "ts{id=1, aggregate=\\"discreteVariance\\", granularity=\\"10d\\"}",
        "limit": 10000,
        "start": 1549336675000,
      },
    ],
  },
  "method": "POST",
  "requestId": "1_11_B",
  "url": "/api/datasources/proxy/6/playground/TestProject/timeseries/synthetic/query",
}
`;

exports[`Datasource Query Given custom queries should generate the correct filtered queries 5`] = `
Object {
  "data": Object {
    "items": Array [
      Object {
        "end": 1549338475000,
        "expression": "ts{externalId=\\"Timeseries1\\"}",
        "limit": 100000,
        "start": 1549336675000,
      },
    ],
  },
  "method": "POST",
  "requestId": "1_11_D",
  "url": "/api/datasources/proxy/6/playground/TestProject/timeseries/synthetic/query",
}
`;

exports[`Datasource Query Given custom queries should return correct datapoints and labels 1`] = `
Object {
  "data": Array [
    Object {
      "datapoints": Array [],
      "target": "Timeseries1",
    },
    Object {
      "datapoints": Array [],
      "target": "Timeseries1",
    },
  ],
}
`;

exports[`Datasource Query Given custom queries with functions should generate the correct filtered queries 1`] = `
Object {
  "data": Object {
    "filter": Object {},
    "limit": 1000,
  },
  "method": "POST",
  "url": "/api/datasources/proxy/6/cogniteapi/TestProject/timeseries/list",
}
`;

exports[`Datasource Query Given custom queries with functions should generate the correct filtered queries 2`] = `
Object {
  "data": Object {
    "items": Array [
      Object {
        "id": 1,
      },
      Object {
        "id": 2,
      },
      Object {
        "id": 3,
      },
    ],
  },
  "method": "POST",
  "url": "/api/datasources/proxy/6/cogniteapi/TestProject/timeseries/byids",
}
`;

exports[`Datasource Query Given custom queries with functions should generate the correct filtered queries 3`] = `
Object {
  "data": Object {
    "items": Array [
      Object {
        "id": 1,
      },
      Object {
        "externalId": "Timeseries1",
      },
      Object {
        "id": 2,
      },
      Object {
        "id": 3,
      },
    ],
  },
  "method": "POST",
  "url": "/api/datasources/proxy/6/cogniteapi/TestProject/timeseries/byids",
}
`;

exports[`Datasource Query Given custom queries with functions should generate the correct filtered queries 4`] = `
Object {
  "data": Object {
    "items": Array [
      Object {
        "end": 1549338475000,
        "expression": "ts{id=1} + pi()",
        "limit": 33333,
        "start": 1549336675000,
      },
      Object {
        "end": 1549338475000,
        "expression": "ts{id=2} + pi()",
        "limit": 33333,
        "start": 1549336675000,
      },
      Object {
        "end": 1549338475000,
        "expression": "ts{id=3} + pi()",
        "limit": 33333,
        "start": 1549336675000,
      },
    ],
  },
  "method": "POST",
  "requestId": "1_14_A",
  "url": "/api/datasources/proxy/6/playground/TestProject/timeseries/synthetic/query",
}
`;

exports[`Datasource Query Given custom queries with functions should generate the correct filtered queries 5`] = `
Object {
  "data": Object {
    "items": Array [
      Object {
        "end": 1549338475000,
        "expression": "ts{id=1} * ts{externalId=\\"Timeseries1\\", aggregate=\\"average\\"} - ts{externalId=\\"Timeseries1\\", aggregate=\\"average\\", granularity=\\"10m\\"}",
        "limit": 3333,
        "start": 1549336675000,
      },
      Object {
        "end": 1549338475000,
        "expression": "ts{id=2} * ts{externalId=\\"Timeseries1\\", aggregate=\\"average\\"} - ts{externalId=\\"Timeseries1\\", aggregate=\\"average\\", granularity=\\"10m\\"}",
        "limit": 3333,
        "start": 1549336675000,
      },
      Object {
        "end": 1549338475000,
        "expression": "ts{id=3} * ts{externalId=\\"Timeseries1\\", aggregate=\\"average\\"} - ts{externalId=\\"Timeseries1\\", aggregate=\\"average\\", granularity=\\"10m\\"}",
        "limit": 3333,
        "start": 1549336675000,
      },
    ],
  },
  "method": "POST",
  "requestId": "1_14_D",
  "url": "/api/datasources/proxy/6/playground/TestProject/timeseries/synthetic/query",
}
`;

exports[`Datasource Query Given custom queries with functions should return correct datapoints and labels 1`] = `
Object {
  "data": Array [
    Object {
      "datapoints": Array [],
      "target": "test timeseriesA value1 + pi()",
    },
    Object {
      "datapoints": Array [],
      "target": "test timeseriesB value1 + pi()",
    },
    Object {
      "datapoints": Array [],
      "target": "test timeseriesC value1 + pi()",
    },
    Object {
      "datapoints": Array [],
      "target": "Timeseries1 * Timeseries1 - Timeseries1",
    },
    Object {
      "datapoints": Array [],
      "target": "Timeseries2 * Timeseries1 - Timeseries1",
    },
    Object {
      "datapoints": Array [],
      "target": "Timeseries3 * Timeseries1 - Timeseries1",
    },
  ],
}
`;
